modules/core/src/main/java/org/apache/ignite/internal/processors/rest/handlers/cluster/GridBaselineCommandHandler.java

```diff
@@ -0,0 +1,201 @@
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toMap;
+import static org.apache.ignite.internal.processors.rest.GridRestCommand.BASELINE_ADD;
+import static org.apache.ignite.internal.processors.rest.GridRestCommand.BASELINE_CURRENT_STATE;
+import static org.apache.ignite.internal.processors.rest.GridRestCommand.BASELINE_REMOVE;
+import static org.apache.ignite.internal.processors.rest.GridRestCommand.BASELINE_SET;
+
+/**
+ *
+ */
+public class GridBaselineCommandHandler extends GridRestCommandHandlerAdapter {
+    /** Supported commands. */
+    private static final Collection<GridRestCommand> SUPPORTED_COMMANDS = U.sealList(BASELINE_CURRENT_STATE,
+        BASELINE_SET, BASELINE_ADD, BASELINE_REMOVE);
+
+    /**
+     * @param ctx Context.
+     */
+    public GridBaselineCommandHandler(GridKernalContext ctx) {
+        super(ctx);
+    }
+
+    /** {@inheritDoc} */
+    @Override public Collection<GridRestCommand> supportedCommands() {
+        return SUPPORTED_COMMANDS;
+    }
+
+    /** {@inheritDoc} */
+    @Override public IgniteInternalFuture<GridRestResponse> handleAsync(GridRestRequest req) {
+        try {
+            GridBaselineCommandResponse baseline = handleAsyncUnsafe(req);
+
+            return new GridFinishedFuture<>(new GridRestResponse(baseline));
+        }
+        catch (IgniteCheckedException e) {
+            if (!X.hasCause(e, VisorClusterGroupEmptyException.class))
+                U.error(log, "Failed to execute task command: " + req, e);
+
+            return new GridFinishedFuture<>(e);
+        }
+        finally {
+            if (log.isDebugEnabled())
+                log.debug("Handled task REST request: " + req);
+        }
+    }
+
+    /**
+     * @return Current baseline.
+     */
+    private Set<BaselineNode> currentBaseLine() {
+        return Optional.ofNullable(ctx.grid().cluster().currentBaselineTopology())
+            .map((Function<Collection<BaselineNode>, Set<BaselineNode>>)HashSet::new)
+            .orElseGet(HashSet::new);
+    }
+
+    /**
+     * Collect baseline and server nodes.
+     *
+     * @return Baseline descriptor.
+     */
+    private GridBaselineCommandResponse collect() {
+        IgniteClusterEx cluster = ctx.grid().cluster();
+
+        Collection<? extends BaselineNode> srvrs = cluster.forServers().nodes();
+
+        return new GridBaselineCommandResponse(cluster.active(), cluster.topologyVersion(), currentBaseLine(), srvrs);
+    }
+
+    /**
+     * Set new baseline.
+     *
+     * @param baselineTop Initial collection of nodes.
+     * @param consistentIds Collection of consistent IDs to set.
+     */
+    private void set(Set<BaselineNode> baselineTop, List<Object> consistentIds) {
+        IgniteClusterEx cluster = ctx.grid().cluster();
+        
+        Map<Object, BaselineNode> srvrs = cluster.forServers().nodes().stream()
+            .collect(toMap(n -> String.valueOf(n.consistentId()), identity()));
```

Comment:
Same question about nullability of consistent ids. "valueOf" returns "null" as a String for empty strings, I think it's not what we want.